libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/ptx/ptx_helper_functions.h

```diff
@@ -26,37 +27,82 @@ _LIBCUDACXX_BEGIN_NAMESPACE_CUDA_PTX
 
 inline _LIBCUDACXX_DEVICE _CUDA_VSTD::uint32_t __as_ptr_smem(const void* __ptr)
 {
-  // Consider adding debug asserts here.
-  return static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__ptr));
+  NV_DISPATCH_TARGET(
+    NV_IS_DEVICE, (
+      // Consider adding debug asserts here.
+      return static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__ptr));
+    ),
+    NV_ANY_TARGET, (
+      // UNREACHABLE
+      _LIBCUDACXX_UNREACHABLE();
+      return 0;
+    )
+  )
 }
 
 inline _LIBCUDACXX_DEVICE _CUDA_VSTD::uint32_t __as_ptr_remote_dsmem(const void* __ptr)
 {
-  // No difference in implementation to __as_ptr_smem.
-  // Consider adding debug asserts here.
-  return static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__ptr));
+  NV_DISPATCH_TARGET(
+    NV_IS_DEVICE, (
+      // No difference in implementation to __as_ptr_smem.
+      // Consider adding debug asserts here.
+      return static_cast<_CUDA_VSTD::uint32_t>(__cvta_generic_to_shared(__ptr));
+    ),
+    NV_ANY_TARGET, (
+      // UNREACHABLE
+      _LIBCUDACXX_UNREACHABLE();
+      return 0;
+    )
+  )
 }
 
 inline _LIBCUDACXX_DEVICE _CUDA_VSTD::uint64_t __as_ptr_gmem(const void* __ptr)
 {
-  // Consider adding debug asserts here.
-  return static_cast<_CUDA_VSTD::uint64_t>(__cvta_generic_to_global(__ptr));
+  NV_DISPATCH_TARGET(
+    NV_IS_DEVICE, (
```

Comment:
Question: do we need NV_DISPATCH_TARGET in device functions as well?

Reply:
I am also curious about that one

Reply:
Yep, headers should be includable in host only TUs.